File: test-suite/src/test/java/io/micronaut/email/docs/MockEmailSender.java
package io.micronaut.email.docs;

import io.micronaut.context.annotation.Requires;
import io.micronaut.core.util.StringUtils;
import io.micronaut.email.test.Mailpit;
import io.micronaut.runtime.http.scope.RequestScope;
import org.jspecify.annotations.NonNull;
import io.micronaut.email.Email;
import io.micronaut.email.EmailException;
import io.micronaut.email.TransactionalEmailSender;
import jakarta.inject.Named;
import jakarta.inject.Singleton;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

@Requires(property = "mock.emailsender", value = StringUtils.TRUE)
@Named("mock")
@Singleton
public class MockEmailSender<I> implements TransactionalEmailSender<I, Email> {

    private final List<Email> emails = new ArrayList<>();
    private final List<Consumer<I>> requests = new ArrayList<>();

    public List<Email> getEmails() {
        return emails;
    }

    public List<Consumer<I>> getRequests() {
        return requests;
    }

    @Override
    @NonNull
    public String getName() {
        return "mock";
    }

    @NonNull
    @Override
    public Email send(@NonNull @NotNull @Valid Email email,
                      @NonNull @NotNull Consumer<I> emailRequest) throws EmailException {
        emails.add(email);
        requests.add(emailRequest);
        return email;
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/email/docs/MockEmailSender.kt
package io.micronaut.email.docs

import io.micronaut.core.annotation.NonNull
import io.micronaut.email.Email
import io.micronaut.email.EmailException
import io.micronaut.email.TransactionalEmailSender
import jakarta.inject.Named
import jakarta.inject.Singleton
import java.util.function.Consumer

@Named("mock")
@Singleton
class MockEmailSender<I, O> : TransactionalEmailSender<I, Email> {

    private val emails: MutableList<Email> = mutableListOf()
    private val requests: MutableList<Consumer<I>> = mutableListOf()

    fun getEmails() = emails
    fun getRequests() = requests

    @NonNull
    override fun getName(): String {
        return "mock"
    }

    @Throws(EmailException::class)
    override fun send(email: Email, emailRequest: Consumer<I>): Email {
        emails.add(email)
        requests.add(emailRequest)
        return email
    }
}